### PR TITLE
API, Core: implement types timestamp_ns and timestamptz_ns

### PR DESCRIPTION
**Pull request by jacobmarble**
Helps #8657

This change is divided into two commits:
- API
- Core

This change adds field `ChronoUnit unit` to `TimestampType`, such that `TimestampType` now represents four specified types:
  - `timestamp` (existing)
  - `timestamptz` (existing)
  - `timestamp_ns` (added in #8683)
  - `timestamptz_ns` (added in #8683)

### CODE REVIEW

**Comment by jacobmarble** [comment]
I've tracked down the source of the four failing "core-tests" and will push a fix tomorrow.

**Comment by jacobmarble** [comment]
Fixing four failing tests caused me to do most of the new type work for Parquet and ORC as well. Please note that the ORC type attributes need an update in the spec, else it isn't possible to correctly convert an ORC timestamp to an Iceberg timestamp.

File: api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java [hunk 1]
```diff
@@ -589,17 +603,17 @@ private static String sanitizeNumber(Number value, String type) {
...
-  private static String sanitizeString(CharSequence value, long now, int today) {
+  private static String sanitizeString(CharSequence value, long nowMillis, int today) {
     try {
       if (DATE.matcher(value).matches()) {
         Literal<Integer> date = Literal.of(value).to(Types.DateType.get());
         return sanitizeDate(date.value(), today);
       } else if (TIMESTAMP.matcher(value).matches()) {
-        Literal<Long> ts = Literal.of(value).to(Types.TimestampType.withoutZone());
-        return sanitizeTimestamp(ts.value(), now);
+        Literal<Long> ts = Literal.of(value).to(Types.TimestampType.nanosWithoutZone());
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java:612]: Do we always want to interpret it as NANOs? Depending on the datatype that's being used this might restrict the range of dates.

File: api/src/main/java/org/apache/iceberg/transforms/Days.java [hunk 2]
```diff
@@ -55,14 +56,14 @@ public boolean satisfiesOrderOf(Transform<?, ?> other) {
     }
 
     if (other instanceof Timestamps) {
-      return Timestamps.DAY.satisfiesOrderOf(other);
+      return ((Timestamps) other).getResultTypeUnit() == ChronoUnit.DAYS
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/transforms/Days.java:59]: Can we cast once and store it in a local variable?

File: api/src/main/java/org/apache/iceberg/transforms/Hours.java [hunk 3]
```diff
@@ -57,15 +58,16 @@ public boolean satisfiesOrderOf(Transform<?, ?> other) {
     }
 
     if (other instanceof Timestamps) {
-      return other == Timestamps.HOUR;
-    } else if (other instanceof Hours
-        || other instanceof Days
-        || other instanceof Months
-        || other instanceof Years) {
-      return true;
+      return ((Timestamps) other).getResultTypeUnit() == ChronoUnit.HOURS
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/transforms/Hours.java:61]: Can we cast once and store it in a local variable?

File: api/src/main/java/org/apache/iceberg/transforms/Months.java [hunk 4]
```diff
@@ -55,14 +57,13 @@ public boolean satisfiesOrderOf(Transform<?, ?> other) {
     }
 
     if (other instanceof Timestamps) {
-      return Timestamps.MONTH.satisfiesOrderOf(other);
+      return ((Timestamps) other).getResultTypeUnit() == ChronoUnit.MONTHS
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/transforms/Months.java:60]: Can we cast once and store it in a local variable?

File: api/src/main/java/org/apache/iceberg/transforms/PartitionSpecVisitor.java [hunk 5]
```diff
@@ -121,17 +121,13 @@ static <R> R visit(Schema schema, PartitionField field, PartitionSpecVisitor<R>
     } else if (transform instanceof Truncate) {
       int width = ((Truncate<?>) transform).width();
       return visitor.truncate(field.fieldId(), sourceName, field.sourceId(), width);
-    } else if (transform == Dates.YEAR
-        || transform == Timestamps.YEAR
-        || transform instanceof Years) {
+    } else if ("year".equalsIgnoreCase(transform.toString())) {
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/transforms/PartitionSpecVisitor.java:124]: These changes seem unrelated to me?

File: api/src/main/java/org/apache/iceberg/transforms/Transforms.java [hunk 6]
```diff
@@ -129,10 +131,14 @@ public static <T> Transform<T, Integer> year(Type type) {
...
-      default:
-        throw new IllegalArgumentException("Cannot partition type " + type + " by year");
+        switch (((TimestampType) type).unit()) {
+          case MICROS:
+            return (Transform<T, Integer>) Timestamps.YEAR_FROM_MICROS;
+          case NANOS:
+            return (Transform<T, Integer>) Timestamps.YEAR_FROM_NANOS;
+        }
     }
+    throw new IllegalArgumentException("Cannot partition type " + type + " by year");
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/transforms/Transforms.java:141]: Nit: I think how it was previously, with the `default` reads better

File: api/src/main/java/org/apache/iceberg/types/Types.java [hunk 7]
```diff
@@ -205,27 +208,56 @@ public String toString() {
...
+    private static final TimestampType INSTANCE_MICROS_WITH_ZONE =
+        new TimestampType(true, ChronoUnit.MICROS);
+    private static final TimestampType INSTANCE_MICROS_WITHOUT_ZONE =
+        new TimestampType(false, ChronoUnit.MICROS);
+    private static final TimestampType INSTANCE_NANOS_WITH_ZONE =
+        new TimestampType(true, ChronoUnit.NANOS);
+    private static final TimestampType INSTANCE_NANOS_WITHOUT_ZONE =
+        new TimestampType(false, ChronoUnit.NANOS);
 
     public static TimestampType withZone() {
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/types/Types.java:221]: Should we mark these as deprecated? And suggest using `microsWithZone()` instead?

File: api/src/main/java/org/apache/iceberg/transforms/Days.java (hunk 2 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/transforms/Days.java]: Fixed in 3b87435b31eac16b86b4118e79cb54fabf367b64

File: api/src/main/java/org/apache/iceberg/transforms/Hours.java (hunk 3 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/transforms/Hours.java]: Fixed in 3b87435b31eac16b86b4118e79cb54fabf367b64

File: api/src/main/java/org/apache/iceberg/transforms/Months.java (hunk 4 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/transforms/Months.java]: Fixed in 3b87435b31eac16b86b4118e79cb54fabf367b64

File: api/src/main/java/org/apache/iceberg/transforms/Transforms.java (hunk 6 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/transforms/Transforms.java]: I agree. Fixed in 7b6e973f4ac0cb2ebd1a595eca0aad8ec2c1760f

As I fixed this, I noticed that exceptions thrown for "unsupported timestamp unit" were inconsistent, so I cleaned that up in c52c3d9e96136eb574e4b1547e4c2afd07814ead

File: api/src/main/java/org/apache/iceberg/types/Types.java (hunk 7 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/types/Types.java:225]: Fixed in d7b28e4b3aaa0610ea49ba15b9f3e5a1862f9fac

File: api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java (hunk 1 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java:618]: Resolved in 1a6cf52c1baa35e1fdd086d8907fbed7873cf317

File: api/src/main/java/org/apache/iceberg/transforms/PartitionSpecVisitor.java (hunk 5 above)
Comment (jacobmarble) [reply api/src/main/java/org/apache/iceberg/transforms/PartitionSpecVisitor.java]: You're right. Fixed in 09c1f2534

File: format/spec.md
Comment (emkornfield) [inline format/spec.md:1]: A few high level questions.  Don't the Avro and Parquet details need to be updated?

Also, are these changes intended to be backdate into spec version 1 and 2 or part of v3?

Comment (jacobmarble) [reply format/spec.md:1]: > Don't the Avro and Parquet details need to be updated?

Yes.

(I'm new to the project.) My original intent here was to create multiple PRs, each with a different scope (API/Core, Avro, ORC, Parquet, Spark/Flink/Hive/etc). However, the cascade of failing tests in this PR suggests that one big change might be necessary. Do you have any feedback on this?

> are these changes intended to be backdate into spec version 1 and 2 or part of v3?

These changes affect v3 only.
https://github.com/apache/iceberg/blob/7c4bdaa3a26f2a8b36f22b1a1257aa11bf136969/format/spec.md?plain=1#L184-L185
https://github.com/apache/iceberg/blob/7c4bdaa3a26f2a8b36f22b1a1257aa11bf136969/format/spec.md?plain=1#L1302

**Comment by jbonofre** [comment]
I'm starting to review. Thanks !

**Comment by rdblue** [comment]
@jacobmarble can you break this into smaller commits? There are a ton of files changed here and I'm concerned about catching problems with such a large PR.

**Comment by jbonofre** [comment]
I agree with @rdblue 

For partition stat we had a PR dedicated for spec. 
As it's not necessary the same reviewers, I think it's good to have one PR for spec (even if it's very small) and PRs for api and impl.

**Comment by jacobmarble** [comment]
> @jacobmarble can you break this into smaller commits? There are a ton of files changed here and I'm concerned about catching problems with such a large PR.

> I agree with @rdblue
> 
> For partition stat we had a PR dedicated for spec. As it's not necessary the same reviewers, I think it's good to have one PR for spec (even if it's very small) and PRs for api and impl.

Happy to split this up. I tried to scope this to API and Core, but the cascade of failing tests led to this.
